`Gem::Requirement#==` by `OrderIndependentComparison` raising `NoMethodError` on RubyGems < 3.2.0

I'm having a problem.

My current problem is

I am trying to run Ruby 2.7.3 (RubyGems vesion is 3.1.6)'s tests (`test/rubygems/test_gem_requirement.rb`) with manually backported Bundler 2.2.20 from Bundler 2.1.4 to find a possibility to fix a security issue related to Bundler. On the RubyGems 3.1.6 (< RubyGems 3.2.0.rc.2), in the following code, `Gem::Requirement::OrderIndependentComparison` is loaded, and `Gem::Requirement#==` is overridden.

test/rubygems/test_gem_requirement.rb

https://github.com/rubygems/rubygems/blob/f88fa556d68988e478c2a7881682871b58e62643/test/rubygems/test_gem_requirement.rb#L31

bundler/lib/bundler/rubygems_ext.rb

https://github.com/rubygems/rubygems/blob/f88fa556d68988e478c2a7881682871b58e62643/bundler/lib/bundler/rubygems_ext.rb#L132-L159

This logic was added by https://github.com/rubygems/rubygems/commit/c33914fe2b8c0ebba1f0b4b72f04152e640254c5 .

Then running the test in ruby/ruby, as `Object.new` (`other` object) doen't have `_requirements_sorted` method, I see the following `NoMethodError` error by the above assertion line in `test/rubygems/test_gem_requirement.rb` and `other._requirements_sorted?` in `bundler/lib/bundler/rubygems_ext.rb`.

```
  2) Error:
TestGemRequirement#test_equals2:
NoMethodError: undefined method `_requirements_sorted?' for #<Object:0x0000562bc2858330>
    /builddir/build/BUILD/ruby-2.7.3/lib/bundler/rubygems_ext.rb:137:in `=='
```

To simply the issue, actually this issue can be reproduced on the latest rubygems repo too.

On the latest rubygems master commit `f88fa556d68988e478c2a7881682871b58e62643`, 

```
$ ruby -v
ruby 3.0.1p64 (2021-04-05 revision 0fb782ee38) [x86_64-linux]

$ rake setup
```

Comment out the condition `unless ...` logic to emulate RubyGems < 3.2.0.

```
$ vi bundler/lib/bundler/rubygems_ext.rb

$ git diff
diff --git a/bundler/lib/bundler/rubygems_ext.rb b/bundler/lib/bundler/rubygems_ext.rb
index c95664965c..cd4306aabf 100644
--- a/bundler/lib/bundler/rubygems_ext.rb
+++ b/bundler/lib/bundler/rubygems_ext.rb
@@ -130,7 +130,7 @@ def to_lock
   end
 
   # comparison is done order independently since rubygems 3.2.0.rc.2
-  unless Gem::Requirement.new("> 1", "< 2") == Gem::Requirement.new("< 2", "> 1")
+  # unless Gem::Requirement.new("> 1", "< 2") == Gem::Requirement.new("< 2", "> 1")
     class Requirement
       module OrderIndependentComparison
         def ==(other)
@@ -156,7 +156,7 @@ def _with_sorted_requirements
 
       prepend OrderIndependentComparison
     end
-  end
+  # end
 
   if Gem::Requirement.new("~> 2.0").hash == Gem::Requirement.new("~> 2.0.0").hash
     class Requirement
```

Then run the target test method.

```
$ ruby -Ilib:test test/rubygems/test_gem_requirement.rb -n /test_equals2/
Loaded suite test/rubygems/test_gem_requirement
Started
E
========================================================================================================
Error: test_equals2(TestGemRequirement): NoMethodError: undefined method `_requirements_sorted?' for #<Object:0x00000000015334b8>
/home/jaruga/var/git/rubygems/bundler/lib/bundler/rubygems_ext.rb:137:in `=='
test/rubygems/test_gem_requirement.rb:31:in `test_equals2'
========================================================================================================

Finished in 0.023809665 seconds.
--------------------------------------------------------------------------------------------------------
1 tests, 11 assertions, 0 failures, 1 errors, 0 pendings, 0 omissions, 0 notifications
0% passed
--------------------------------------------------------------------------------------------------------
42.00 tests/s, 462.00 assertions/s
Coverage report generated for Unit Tests to /home/jaruga/var/git/rubygems/coverage. 1085 / 3578 LOC (30.32%) covered.
```

My question is 

The test `test/rubygems/test_gem_requirement.rb` `refute_equal req("= 1.2"), Object.new ` could be modified or `rubygems/bundler/lib/bundler/rubygems_ext.rb` `Gem::Requirement::OrderIndependentComparison` `==` method could be modified?

This issue is related to:

  - [ ] Network problems
  - [ ] Installing a library
  - [ ] Publishing a library
  - [ ] The command line `gem`
  - [X] Other

Here are my current environment details:

```
$ gem env version
3.1.6
```

I will abide by the [code of conduct](https://github.com/rubygems/rubygems/blob/master/CODE_OF_CONDUCT.md).

OK, not to remove the overriden `==` method. but the suggested line `return unless Gem::Requirement == other` might not be correct. I was able to pass the test adding the `return unless other.class == Gem::Requirement` guard line at the first line of the `==` method.

My environment on `ruby/ruby` is here again.

```
<mock-chroot> sh-5.1$ ruby -v
ruby 2.7.3p183 (2021-04-05 revision 6847ee089d) [x86_64-linux]
<mock-chroot> sh-5.1$ gem env version
3.1.6
<mock-chroot> sh-5.1$ gem -v
3.1.6
<mock-chroot> sh-5.1$ bundle -v
Bundler version 2.2.20
```

With your suggested patch,

```
<mock-chroot> sh-5.1$ diff -u lib/bundler/rubygems_ext.rb.org lib/bundler/rubygems_ext.rb
--- lib/bundler/rubygems_ext.rb.org	2021-06-11 13:25:04.000000000 +0200
+++ lib/bundler/rubygems_ext.rb	2021-07-01 23:09:55.722897998 +0200
@@ -134,6 +134,7 @@
     class Requirement
       module OrderIndependentComparison
         def ==(other)
+          return unless Gem::Requirement == other
           if _requirements_sorted? && other._requirements_sorted?
             super
           else
```

The test still fails.

```
<mock-chroot> sh-5.1$ make test-all V=1 TESTS="-v test/rubygems/test_gem_requirement.rb -n TestGemRequirement#test_equals2"
: > revision.tmp
./tool/ifchange "--timestamp=.revision.time" "./revision.h" revision.tmp
./revision.h unchanged
# rm -f revision.h
# : > revision.h || : > revision.h
exec ./miniruby -I./lib -I. -I.ext/common  ./tool/runruby.rb --extout=.ext  -- --disable-gems "./test/runner.rb" --ruby="./miniruby -I./lib -I. -I.ext/common  ./tool/runruby.rb --extout=.ext  -- --disable-gems" --excludes-dir=./test/excludes --name=!/memory_leak/  -v test/rubygems/test_gem_requirement.rb -n TestGemRequirement#test_equals2
Run options: "--ruby=./miniruby -I./lib -I. -I.ext/common  ./tool/runruby.rb --extout=.ext  -- --disable-gems" --excludes-dir=./test/excludes --name=!/memory_leak/ -v -n TestGemRequirement#test_equals2

# Running tests:

[1/0] TestGemRequirement#test_equals2 = 0.02 s

  1) Failure:
TestGemRequirement#test_equals2 [/builddir/build/BUILD/ruby-2.7.3/test/rubygems/test_gem_requirement.rb:17]:
No visible difference in the Gem::Requirement#inspect output.
You should look at the implementation of #== on Gem::Requirement or its members.
Gem::Requirement.new(["= 1.2"])

Finished tests in 0.038007s, 26.3107 tests/s, 26.3107 assertions/s.
1 tests, 1 assertions, 1 failures, 0 errors, 0 skips

ruby -v: ruby 2.7.3p183 (2021-04-05 revision 6847ee089d) [x86_64-linux]
make: *** [uncommon.mk:783: yes-test-all] Error 1
```

Here is the patch I suggest.

```
<mock-chroot> sh-5.1$ diff -u lib/bundler/rubygems_ext.rb.org lib/bundler/rubygems_ext.rb
--- lib/bundler/rubygems_ext.rb.org	2021-06-11 13:25:04.000000000 +0200
+++ lib/bundler/rubygems_ext.rb	2021-07-01 22:59:33.123926006 +0200
@@ -134,6 +134,7 @@
     class Requirement
       module OrderIndependentComparison
         def ==(other)
+          return unless other.class == Gem::Requirement
           if _requirements_sorted? && other._requirements_sorted?
             super
           else
```

The test passed. 😄 

```
<mock-chroot> sh-5.1$ make test-all V=1 TESTS="-v test/rubygems/test_gem_requirement.rb -n TestGemRequirement#test_equals2"
: > revision.tmp
./tool/ifchange "--timestamp=.revision.time" "./revision.h" revision.tmp
./revision.h unchanged
# rm -f revision.h
# : > revision.h || : > revision.h
exec ./miniruby -I./lib -I. -I.ext/common  ./tool/runruby.rb --extout=.ext  -- --disable-gems "./test/runner.rb" --ruby="./miniruby -I./lib -I. -I.ext/common  ./tool/runruby.rb --extout=.ext  -- --disable-gems" --excludes-dir=./test/excludes --name=!/memory_leak/  -v test/rubygems/test_gem_requirement.rb -n TestGemRequirement#test_equals2
Run options: "--ruby=./miniruby -I./lib -I. -I.ext/common  ./tool/runruby.rb --extout=.ext  -- --disable-gems" --excludes-dir=./test/excludes --name=!/memory_leak/ -v -n TestGemRequirement#test_equals2

# Running tests:

[1/0] TestGemRequirement#test_equals2 = 0.02 s
Finished tests in 0.034637s, 28.8708 tests/s, 288.7084 assertions/s.
1 tests, 10 assertions, 0 failures, 0 errors, 0 skips

ruby -v: ruby 2.7.3p183 (2021-04-05 revision 6847ee089d) [x86_64-linux]
```

--- a/.github/workflows/older-rubygems-bundler.yml
+++ b/.github/workflows/older-rubygems-bundler.yml
@@ -61,4 +61,8 @@ jobs:
         run: |
           bin/rake spec:all
         working-directory: ./bundler
+      - name: Run Rubygems Requirement tests against local bundler, to make sure bundler monkeypatches preserve the behaviour
+        run: |
+          ruby -I../../lib:lib:test test/rubygems/test_gem_requirement.rb
+        working-directory: ./bundler/tmp/rubygems
     timeout-minutes: 60

--- a/bundler/lib/bundler/rubygems_ext.rb
+++ b/bundler/lib/bundler/rubygems_ext.rb
@@ -134,6 +134,8 @@ module Gem
     class Requirement
       module OrderIndependentComparison
         def ==(other)
+          return unless Gem::Requirement === other
+
           if _requirements_sorted? && other._requirements_sorted?
             super
           else

